crates/node/src/service/l1.rs

```diff
@@ -1,87 +1,214 @@
+                > = Box::new(client);
+                Some(Arc::new(client_converted))
             } else {
                 anyhow::bail!(
-                    "No Ethereum endpoint provided. You need to provide one using --l1-endpoint <RPC URL> in order to verify the synced state or disable the l1 watcher using --no-l1-sync."
+                    "No Starknet endpoint provided. Use --l1-endpoint <RPC URL> or disable with --no-l1-sync."
                 );
             }
         } else {
             None
         };
 
-        // Note: gas price should be synced in case the madara is running in sequencer mode,
-        // we haven't set any fix price for the gas, hence gas price should be none
+        Self::create_service(config, db, settlement_client, l1_gas_provider, chain_id, authority, devnet, mempool).await
+    }
+}
+
+// Shared implementation for both services
+impl<C: Clone, P: Clone> L1SyncService<C, P> {
+    #[allow(clippy::too_many_arguments)]
+    async fn create_service(
+        config: &L1SyncParams,
+        db: &DatabaseService,
+        settlement_client: Option<Arc<Box<dyn ClientTrait<Config = C, Provider = P>>>>,
+        l1_gas_provider: GasPriceProvider,
+        chain_id: ChainId,
+        authority: bool,
+        devnet: bool,
+        mempool: Arc<Mempool>,
+    ) -> anyhow::Result<Self> {
         let gas_price_sync_enabled =
             authority && !devnet && (config.gas_price.is_none() || config.blob_gas_price.is_none());
         let gas_price_poll = config.gas_price_poll;
 
         if gas_price_sync_enabled {
-            let eth_client = eth_client
-                .clone()
-                .context("L1 gas prices require the ethereum service to be enabled. Either disable gas prices syncing using `--gas-price 0`, or disable L1 sync using the `--no-l1-sync` argument.")?;
-            // running at-least once before the block production service
+            let settlement_client =
+                settlement_client.clone().context("L1 gas prices require the service to be enabled...")?;
             tracing::info!("⏳ Getting initial L1 gas prices");
-            mc_eth::l1_gas_price::gas_price_worker_once(&eth_client, &l1_gas_provider, gas_price_poll)
+            mc_settlement_client::gas_price::gas_price_worker_once(settlement_client, &l1_gas_provider, gas_price_poll)
                 .await
-                .context("Getting initial ethereum gas prices")?;
+                .context("Getting initial gas prices")?;
         }
 
         Ok(Self {
             db_backend: Arc::clone(db.backend()),
-            eth_client,
+            settlement_client,
             l1_gas_provider,
             chain_id,
             gas_price_sync_disabled: !gas_price_sync_enabled,
             gas_price_poll,
             mempool,
         })
     }
+
+    // Factory method to create the appropriate service
+    #[allow(clippy::too_many_arguments)]
```

Comment:
should we make it a struct?

Reply:
I tried but lifetime specifier ka issue aa rha h. will try it in next commit.